### PR TITLE
chore: prep npm packages for use with Cypress v10

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes https://github.com/cypress-io/cypress/issues/21944

### User facing changelog
Publish Cypress v10 compatible npm packages

### Additional details
Removes `"private": true` from the npm packages that we had disabled and fixes a few bugs in our implementation that would have caused issues. Notable changes:

- Replacing the require path for `@cypress/webpack-batteries-included-preprocessor` in webpack-dev-server from `__dirname` to `cypressBinaryRoot`. Relying on `__dirname` only works when sourcing from within the binary (or sourcing from the packages installed node_modules).
- Revert `find-up` to 5.0.0 as >6.0.0 is ESM only. We were relying on `local-pkg` to import it which uses `require`. Removed both packages.
- Restoed webpack-dev-server changelog that got removed in a previous merge
- Fixed types for when using the dev-server function signature

The scope for the original issue became very bloated so I'm not ticking all the boxes from that issue (such as linting non-used deps or releasing as `RC`). Not sure why we came to the decision to publish them as RC but I thoroughly tested all the packages locally and I'm confident we can go forward with a normal breaking change.

I ran the `npm-release.js` script locally and the new versions are as expected:
```
node ./scripts/npm-release.js 
Found the following public packages: create-cypress-tests, @cypress/schematic, @cypress/eslint-plugin-dev, @cypress/mount-utils, @cypress/react, @cypress/vite-dev-server, @cypress/vue, @cypress/vue2, @cypress/webpack-batteries-included-preprocessor, @cypress/webpack-dev-server, @cypress/webpack-preprocessor

Finding package versions...

Cypress binary: 10.0.3

create-cypress-tests
Current version: 1.3.0
Next version: 2.0.0

@cypress/schematic
Current version: 2.0.0
Next version: N/A

@cypress/eslint-plugin-dev
Current version: 5.3.0
Next version: N/A

@cypress/mount-utils
Current version: 1.0.2
Next version: 2.0.0

@cypress/react
Current version: 5.12.5
Next version: 6.0.0

@cypress/vite-dev-server
Current version: 2.2.3
Next version: 3.0.0

@cypress/vue
Current version: 3.1.2
Next version: 4.0.0

@cypress/vue2
Current version: N/A
Next version: 1.0.0

@cypress/webpack-batteries-included-preprocessor
Current version: 2.2.3
Next version: N/A

@cypress/webpack-dev-server
Current version: 1.8.4
Next version: 2.0.0

@cypress/webpack-preprocessor
Current version: 5.12.0
Next version: N/A

Found a new release for the following packages: create-cypress-tests, @cypress/mount-utils, @cypress/react, @cypress/vite-dev-server, @cypress/vue, @cypress/vue2, @cypress/webpack-dev-server
```

We should merge with care. I included two commits, one with all the changes to the files in `npm/` with the `BREAKING CHANGE` commit message and another with the rest of the changes. I didn't want to include a `BREAKING CHANGE` commit message that touched files outside of npm. When merging, I'd like to not squash.

Need to follow up with DX about documentation. There is a [section](https://docs.cypress.io/guides/component-testing/component-framework-configuration#Custom-Dev-Server) about custom-dev-servers, gotta check if this needs to be expanded.

### Steps to test
I tested the packages by running `yarn pack` and installing the generated `.tgz`. I added a few system tests for the dev-servers to verify the use of the function signature.

### How has the user experience changed?
na

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by BlueWinds — CHANGES_REQUESTED** [review]

File: npm/webpack-dev-server/src/makeWebpackConfig.ts
Comment (ZachJW34) [reply npm/webpack-dev-server/src/makeWebpackConfig.ts:104]: Yeah I was reverting to what was there before, but I confirmed it doesn't work. Turns out the TS compilation will convert `import` to a require which is not what we want (see [here](https://github.com/TypeStrong/ts-node/discussions/1290)). There are some workarounds, I'll check both webpack and vite since we do similar things here

Comment (lmiller1990) [reply npm/webpack-dev-server/src/makeWebpackConfig.ts:104]: Going over this PR now -- we can sync on this if needed, I've been dealing with module stuff lately. Not saying I have a perfect solution, but we can explore some things.

If we change the `tsconfig` to `"module": "esnext" it will **not** generate `require`, but maintain the original `import`.

Actually, thinking about it, we have the **exact same problem** in `packages/server`, where we need to grab their `cypress.config`, which can be:

- `"type": "module"` with `js`
- `"type": "CommonJS"` (the default) with `.mjs`
- `.ts` with "module": "CommonJS" in `tsconfig.json`
- `.ts` with "module": "esnext" in `tsconfig.json`

It's also possible to have `.cjs` and, believe it or not, `.mjs` and `.mts`. 

I think we probably want to have one single, uniform way to handle "all the things" if possible. The only difference, I guess, would be that the binary ships `ts-node`, so we know we can use that, but `npm/webpack-dev-server` cannot rely on `ts-node` been installed.

File: npm/vite-dev-server/cypress/e2e/vite-dev-server.cy.ts
Comment (lmiller1990) [inline npm/vite-dev-server/cypress/e2e/vite-dev-server.cy.ts:51]: What does this test actually prove works?

**Review by lmiller1990 — COMMENTED** [review]
Left some general comments - I'm wondering if we don't revert the module loading changes you made and do them separately (eg, this ticket was just about "release the things" not necessarily fixing bugs?

File: npm/vite-dev-server/cypress/e2e/vite-dev-server.cy.ts
Comment (ZachJW34) [reply npm/vite-dev-server/cypress/e2e/vite-dev-server.cy.ts:51]: The system test is hooked up the use `import { devServer } from '@cypress/vite-dev-server` via the `file:` package alias, so it's a test to prove that importing this as a standalone package and using the function signature works.

File: npm/vite-dev-server/package.json
Comment (ZachJW34) [reply npm/vite-dev-server/package.json:19]: One thing to consider here is that these packages are always going to be consumed through our loader process since they are imported in the config (which is processed in the child process we spin up). Not quite sure of all the implications of this, but if we always transpile the config to commonjs then it might factor into how we decide to bundle these packages.

I'm seeing the need for a clearly defined tech-brief just talking through all of this!

Comment (lmiller1990) [reply npm/vite-dev-server/package.json:19]: Ah yes, we'd need to change our development workflow to transpire this on the fly if we wanted to write pure ESM (probably not a big deal, but more work than really makes sense now).

That said, unless we eventually do commit to it, we will likely be stuck on ESM forever (like most of the ecosystem). We finally got the module system we all dreamed off, but it came 10 years too late :(

**Comment by ZachJW34** [comment]
@lmiller1990 @BlueWinds I decided to revert the changes made to how the webpack config was being loaded. I wrote a decent amount of test cases to cover all the variations of js + cjs + ts + esm and they were pros and cons for each approach. The way it is wired up now is actually the best for our use case of providing good support for the most common cases (js + ts + cjs).

I've learned that there is a workaround for esm and ts edge-cases which is to import the webpack config into the cypress config i.e. not rely on `@cypress/webpack-dev-server`. For example, if the cypress config is loaded as `type=module` and the webpack config is imported there it is processed with the right context (same with typescript).

We have some work to do for handling esm, and I'm going to push up my test cases and link them to a new issue where we can properly work through adding first class support for esm. I feel better having this issue just be related to publishing the packages rather than solve the nightmare that is support for cjs + esm.

**Review by lmiller1990 — APPROVED** [review]
Cool, good call on not doing too deep in the module stuff. #21939 seems to be working pretty well, so we can expand on that approach moving forward. Let's work together on a brief to figure this out once and for all.

Windows CI is fixed in develop, you might need to wait for develop to merge into master before this is able to be merged.

**Review by BlueWinds — APPROVED** [review]

**Comment by ZachJW34** [comment]
Going to merge today, there is a CI issue that I'm merging to develop and will get merged into master soon: https://github.com/cypress-io/cypress/pull/22289

**Comment by ZachJW34** [comment]
Percy-finalize having an issue, all test are passing so going to merge.

**Pull request merged**
